modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/DistributionZoneManager.java

```diff
@@ -806,4 +866,230 @@ private static Throwable unwrapDistributionZoneExceptionRecursively(Throwable e,
+
+            Set<String> dataNodesFromMetaStorage = fromBytes(values.get(zoneDataNodesKey(zoneId)).value());
+
+            long scaleUpTriggerRevision = bytesToLong(values.get(zoneScaleUpChangeTriggerKey(zoneId)).value());
+
+            long scaleDownTriggerRevision = bytesToLong(values.get(zoneScaleDownChangeTriggerKey(zoneId)).value());
+
+            if (revision <= scaleUpTriggerRevision) {
+                return completedFuture(false);
+            }
+
+            ReentrantLock lockForZone = zoneState.lock();
+
+            lockForZone.lock();
+
+            Set<String> deltaToAdd;
+
+            try {
+                deltaToAdd = new HashSet<>(zoneState.nodesToAdd());
+            } finally {
+                lockForZone.unlock();
+            }
+
+            Set<String> newDataNodes = new HashSet<>(dataNodesFromMetaStorage);
+
+            newDataNodes.addAll(deltaToAdd);
+
+            Update dataNodesAndTriggerKeyUpd = updateDataNodesAndScaleUpTriggerKey(zoneId, revision, toBytes(newDataNodes));
+
+            If iif = If.iif(
+                    triggerScaleUpScaleDownKeysCondition(scaleUpTriggerRevision, scaleDownTriggerRevision, zoneId),
+                    dataNodesAndTriggerKeyUpd,
+                    ops().yield(false)
+            );
+
+            return metaStorageManager.invoke(iif).thenApply(StatementResult::getAsBoolean).thenCompose(invokeResult -> {
+                if (invokeResult) {
+                    lockForZone.lock();
+                    try {
+                        zoneState.nodesToAdd().clear();
+                    } finally {
+                        lockForZone.unlock();
+                    }
+                } else {
+                    return saveDataNodesToMetaStorageOnScaleUp(zoneId, revision);
+                }
+
+                return completedFuture(invokeResult);
+            });
+        }).handle((v, e) -> {
+            lockForTimers.unlock();
+
+            if (e != null) {
+                return CompletableFuture.<Boolean>failedFuture(e);
+            }
+
+            return completedFuture(v);
+        }).thenCompose(Function.identity());
+    }
+
+    private class ZoneState {
+        private Timer scaleUp;
+
+        private Timer scaleDown;
+
+        private final ReentrantLock lockDeltas;
+
+        private final ReentrantLock lockForTimers;
+
+        private final Set<String> nodesToAdd;
+
+        private final Set<String> nodesToRemove;
+
+        ZoneState() {
+            this.lockDeltas = new ReentrantLock();
+            this.lockForTimers = new ReentrantLock();
+            this.nodesToAdd = new HashSet<>();
+            this.nodesToRemove = new HashSet<>();
+        }
+
+        private void rescheduleScaleUp(long delay, Runnable runnable) {
+            if (scaleUp == null) {
+                scaleUp = new Timer(runnable);
+            }
+
+            scaleUp.reschedule(delay);
+        }
+
+        private void rescheduleOngoingScaleUp(long delay) {
```

Comment:
Provided logic with (ongoing)rescheduling is error-prone and cumbersome, please rethink and reimplement it. There are 3 possible states for timers:
1. no scheduled task. 
2. task is scheduled but not fired yet.
3. task is executing right now.

In case of 1 and 3 we should schedule new task, so it's possible for current timer to execute one right now and have one more in a pending queue.
In case of option 2 we should reschedule the task. 
Aforementioned scheduling and rescheduling should be thread-safe.